### PR TITLE
config: libintl/libiconv for gettext() detection

### PR DESCRIPTION
**Pull request by rkitover**
Detect in autoconf whether `-lintl` and possibly `-liconv` are necessary
for translation functions like `gettext()`.

The actual autoconf code is just:

```
AM_ICONV
AM_GNU_GETTEXT([external])
LIBS="$LIBS $LTLIBINTL $LTLIBICONV"
```

References:

https://www.gnu.org/software/gettext/manual/html_node/AM_005fGNU_005fGETTEXT.html
https://www.gnu.org/software/gettext/manual/html_node/AM_005fICONV.html

The reason to check for libiconv and add it separately is that this is
sometimes necessary if people are e.g. linking statically.

The `config/*.m4` and `config/config.rpath` files were added by running
`gettextize` and removing everything else.

Tested on alpine and gentoo.

@behlendorf this fixes building on alpine without adding `-lintl` to
`LIBS` explicitly, as promised.

Signed-off-by: Rafael Kitover <rkitover@gmail.com>

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by rkitover** [comment]
I got some unrelated errors building on alpine with kernel `4.20.11`:

```
/home/rkitover/src/zfs/module/spl/spl-condvar.c: In function 'spl_io_schedule_timeout':
/home/rkitover/src/zfs/module/spl/spl-condvar.c:169:2: error: implicit declaration of function 'init_timer'; did you mean 'init_timers'? [-Werror=implicit-function-declaration]                                   
  init_timer(&timer);                                                                                                                                                                                                ^~~~~~~~~~                                                                                                                                                                                                         init_timers                                                                                                                                                                                                      /home/rkitover/src/zfs/module/spl/spl-condvar.c:170:2: error: implicit declaration of function 'setup_timer'; did you mean 'del_timer'? [-Werror=implicit-function-declaration]                                      setup_timer(&timer, __cv_wakeup, (unsigned long)current);                                                                                                                                                          ^~~~~~~~~~~                                                                                                                                                                                                        del_timer                                                                                    
```

and

```
/home/rkitover/src/zfs/module/spl/spl-kstat.c: In function 'proc_kstat_open':
/home/rkitover/src/zfs/module/spl/spl-kstat.c:34:22: error: implicit declaration of function 'PDE' [-Werror=implicit-function-declaration]                                                                         
 #define PDE_DATA(x) (PDE(x)->data)                                                                                                                                                                                                      ^~~                                                                                                                                                                                          /home/rkitover/src/zfs/module/spl/spl-kstat.c:488:15: note: in expansion of macro 'PDE_DATA'                                                                                                                         f->private = PDE_DATA(inode);                                                                                                                                                                                                   ^~~~~~~~                                                                                                                                                                                            /home/rkitover/src/zfs/module/spl/spl-kstat.c:34:28: error: invalid type argument of '->' (have 'int')                                                                                                              #define PDE_DATA(x) (PDE(x)->data)                                                                                                                                                                                
                            ^~
/home/rkitover/src/zfs/module/spl/spl-kstat.c:488:15: note: in expansion of macro 'PDE_DATA'
  f->private = PDE_DATA(inode);
               ^~~~~~~~
```

I could also look into this if you like.

File: config/gettext.m4 [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+  dnl - Invoke AM_ICONV_LINKFLAGS_BODY here, outside any 'if'.
+  dnl - Control the expansions in more detail using AC_PROVIDE_IFELSE.
+  dnl Since AC_PROVIDE_IFELSE is only in autoconf >= 2.52 and not
+  dnl documented, we avoid it.
+  ifelse(gt_included_intl, yes, , [
+    AC_REQUIRE([AM_ICONV_LINKFLAGS_BODY])
+  ])
+
+  dnl Sometimes, on Mac OS X, libintl requires linking with CoreFoundation.
+  gt_INTL_MACOSX
```
Comment (behlendorf) [inline config/gettext.m4]: I noticed the following new error when running `./configure`.

```
./configure: line 35607: gt_INTL_MACOSX: command not found
```

**Comment by behlendorf** [comment]
> I got some unrelated errors building on alpine with kernel 4.20.11:
> I could also look into this if you like.

That would be great.  I believe the second one may already be addressed in master by 9a65234c8bbfe4e394d32172439ae0266ac59c8d,but if not let's get them fixed.

File: config/gettext.m4 (hunk 1 above)
Comment (rkitover) [reply config/gettext.m4]: Sounds like missing macros, I think I may have included too much from gettextize too, I will see if I optimize this somewhat.

What system were you running on?

Comment (behlendorf) [reply config/gettext.m4]: Thanks.  I saw this on Ubuntu 18.04 and CentOS 7 but I suspect it'll be easy to reproduce on most systems.

Comment (rkitover) [reply config/gettext.m4]: I pushed a commit where I ran gettextize from git gettext and it installed the missing `intlmacosx.m4` file. Also fixed a problem with the `config.rpath` by making it an empty file.

**Comment by rkitover** [comment]
This is from the fedora 29 slave:

```
configure: error: 
	*** Kernel built with CONFIG_DEBUG_LOCK_ALLOC which is incompatible
	*** with the CDDL license and will prevent the module linking stage
	*** from succeeding.  You must rebuild your kernel without this
	*** option enabled.
```

is this going to be an issue for people on fedora?

**Comment by rlaager** [comment]
I think that's unrelated to your changes. I saw the same error on a PR of mine that changed only man pages.

**Comment by rkitover** [comment]
Indeed, but if the default kernel in fedora is no longer compatible with zol that's going to be a serious annoyance for people, we should follow this up with fedora and see if they really need these options.

**Comment by behlendorf** [comment]
Sorry about the Fedora build issue, I'll see about resolving it but you can safely ignore it for the moment.  The CI builder is installing the Fedora debug kernel and not the default kernel.  The default kernel works correctly with ZoL.

**Pull request merged**